src/db_impl.cc

```diff
@@ -837,6 +837,50 @@ Status TitanDBImpl::DeleteFilesInRanges(ColumnFamilyHandle* column_family,
   return s;
 }
 
+void TitanDBImpl::MaybeScheduleRangeMerge(
+    const std::vector<std::shared_ptr<BlobFileMeta>>& files,
+    int max_sorted_runs) {
+  mutex_.AssertHeld();
+  if (files.empty()) return;
+
+  // store and sort both ends of blob files to count sorted runs
+  std::vector<std::pair<BlobFileMeta*, bool /* is smallest end? */>> blob_ends;
+  for (const auto& file : files) {
+    blob_ends.emplace_back(std::make_pair(file.get(), true));
+    blob_ends.emplace_back(std::make_pair(file.get(), false));
+  }
+  auto blob_ends_cmp = [](const std::pair<BlobFileMeta*, bool>& end1,
+                          const std::pair<BlobFileMeta*, bool>& end2) {
+    std::string key1 = std::move(end1.second ? end1.first->smallest_key()
```

Comment:
you cannot move the content of smallest_key() or largest_key(), or the next call of comparison on the file will find the key to be empty. Just use reference
```
std::string& key1 = end1.second ? end.first->smallest_key() : end->first->largest_key();
```

Reply:
> you cannot move the content of smallest_key() or largest_key(), or the next call of comparison on the file will find the key to be empty. Just use reference
> 
> ```
> std::string& key1 = end1.second ? end.first->smallest_key() : end->first->largest_key();
> ```

ok. But smallest_key() returned a new constructed string, I thought it won't affect filemeta?